### PR TITLE
Vulkan: handle probe ssbo command

### PR DESCRIPTION
**Pull request by jaebaek**
Related to #32

### CODE REVIEW

File: src/engine.h [hunk 1]
```diff
@@ -33,6 +33,30 @@ enum class PipelineType : uint8_t {
...
 
+enum class ResourceInfoType : uint8_t {
+  kBuffer = 0,
+  kImage,
+};
+
+struct ResourceInfo {
+  ResourceInfoType type;
+
+  union {
```
Comment (dj2) [inline src/engine.h]: Let's not use a union here.

File: src/verifier.cc [hunk 2]
```diff
@@ -15,15 +15,143 @@
...
+}
+
+template <typename T>
+Result CheckIntValue(ProbeSSBOCommand::Comparator comp,
+                     const uint8_t* memory,
+                     const std::vector<Value>& values) {
+  const T* ptr = reinterpret_cast<const T*>(memory);
+  switch (comp) {
+    case ProbeSSBOCommand::Comparator::kEqual:
+      for (auto v : values) {
```
[the remark this answers is not in the document]
Comment (dj2) [reply src/verifier.cc]: Move the for() outside the switch, copy the value into an const T val = static_cast<T>(v.AsUint64); should make this easier to read

File: src/verifier.cc [hunk 3]
```diff
@@ -15,15 +15,143 @@
...
+        if (*ptr != static_cast<T>(v.AsUint64()))
+          return Result("Verifier::ProbeSSBO failed");
+        ++ptr;
+      }
+      break;
+    case ProbeSSBOCommand::Comparator::kNotEqual:
+      for (auto v : values) {
+        if (*ptr == static_cast<T>(v.AsUint64()))
+          return Result("Verifier::ProbeSSBO failed");
```
Comment (dj2) [inline src/verifier.cc]: These results should have more context:

Result("Verifier failed: " + std::to_string(*ptr) + " != " + std::to_string(val));

File: src/verifier.cc [hunk 4]
```diff
@@ -15,15 +15,143 @@
...
+template <typename T>
+Result CheckFloatValue(ProbeSSBOCommand::Comparator comp,
+                       const uint8_t* memory,
+                       const std::vector<Value>& values) {
+  const T* ptr = reinterpret_cast<const T*>(memory);
+  switch (comp) {
+    case ProbeSSBOCommand::Comparator::kEqual:
+      for (auto v : values) {
+        if (*ptr != static_cast<T>(v.AsDouble()))
+          return Result("Verifier::ProbeSSBO failed");
```
Comment (dj2) [inline src/verifier.cc]: Same comments as above

File: src/verifier.cc [hunk 5]
```diff
@@ -15,15 +15,143 @@
...
+      break;
+    default:
+      return Result("Verifier::ProbeSSBO unknown comparator");
+  }
+
+  return {};
+}
+
+template <typename T>
+Result CheckFloatValue(ProbeSSBOCommand::Comparator comp,
```
Comment (dj2) [inline src/verifier.cc]: This is almost identical to the above they can probably be combined just have to change the
static_cast<T>(v.AsDouble) to something like static_cast<T>(v.IsInteger() ? v.AsUint64() : v.AsDouble()).

File: src/verifier.cc [hunk 6]
```diff
@@ -73,11 +201,15 @@ Result Verifier::Probe(const ProbeCommand* command,
     const uint8_t* p = ptr + stride * frame_width * (j + y) + stride * x;
     for (uint32_t i = 0; i < width; ++i) {
       // TODO(jaebaek): Get actual pixel values based on frame buffer formats.
-      if (!IsFloatPixelEqualInt(command->GetR(), p[stride * i]) ||
-          !IsFloatPixelEqualInt(command->GetG(), p[stride * i + 1]) ||
-          !IsFloatPixelEqualInt(command->GetB(), p[stride * i + 2]) ||
+      if (!IsFuzzyEqual(static_cast<double>(command->GetR()),
+                        static_cast<double>(p[stride * i]) / 255.0) ||
```
Comment (dj2) [inline src/verifier.cc]: Why move the / 255.0 out here instead of leaving it in the IsFuzzyEqual?

File: src/verifier.cc [hunk 7]
```diff
@@ -113,8 +245,43 @@ Result Verifier::Probe(const ProbeCommand* command,
   return {};
 }
 
-Result Verifier::ProbeSSBO(const ProbeSSBOCommand*) {
-  return Result("Verifier::ProbeSSBO Not Implemented");
+Result Verifier::ProbeSSBO(const ProbeSSBOCommand* command,
+                           size_t size,
```
Comment (dj2) [inline src/verifier.cc]: Instead of passing a size and a const void*, passing a std::vector<uint8_t>* That will give you both data and size and it will get auto-cleanedup

File: src/vulkan/compute_pipeline.h [hunk 8]
```diff
@@ -37,6 +37,9 @@ class ComputePipeline : public Pipeline {
 
   Result Compute(uint32_t x, uint32_t y, uint32_t z);
 
+  // Pipeline
```
[the remark this answers is not in the document]
Comment (dj2) [reply src/vulkan/compute_pipeline.h]: This is chromium style, put the name of the class you're overriding as a comment before the overrides.

File: src/vulkan/descriptor.h [hunk 9]
```diff
@@ -87,6 +90,8 @@ class Descriptor {
 
   virtual Result UpdateDescriptorSet(VkDescriptorSet descriptor_set) = 0;
   virtual void SendDataToGPUIfNeeded(VkCommandBuffer command) = 0;
```
[the remark this answers is not in the document]
Comment (dj2) [reply src/vulkan/descriptor.h]: So, this was me, heh. I find Host/Device confusing so asked for CPU/GPU to make it clear where the memory was/is.

File: src/vulkan/image.cc [hunk 10]
```diff
@@ -122,9 +122,9 @@ void Image::Shutdown() {
   memory_ = VK_NULL_HANDLE;
 }
 
-Result Image::CopyToHost(VkCommandBuffer command) {
+void Image::CopyToCPU(VkCommandBuffer command) {
   if (is_image_host_accessible_)
-    return {};
+    return;
```
Comment (dj2) [inline src/vulkan/image.cc]: Why isn't this an error if we can't access the image, something has gone wrong, no?

File: src/vulkan/graphics_pipeline.cc [hunk 11]
```diff
@@ -514,9 +514,7 @@ Result GraphicsPipeline::ProcessCommands() {
   ActivateRenderPassIfNeeded();
   DeactivateRenderPassIfNeeded();
 
-  r = frame_->CopyColorImageToHost(command_->GetCommandBuffer());
-  if (!r.IsSuccess())
-    return r;
+  frame_->CopyColorImageToHost(command_->GetCommandBuffer());
```
Comment (dj2) [inline src/vulkan/graphics_pipeline.cc]: What happens if the copy fails?

File: src/vulkan/graphics_pipeline.cc [hunk 12]
```diff
@@ -528,10 +526,6 @@ Result GraphicsPipeline::ProcessCommands() {
 void GraphicsPipeline::Shutdown() {
   DeactivateRenderPassIfNeeded();
 
-  Result r = command_->End();
```
Comment (dj2) [inline src/vulkan/graphics_pipeline.cc]: What happens to any commands we've Begin, but not End'd?

File: src/vulkan/image.cc [hunk 13]
```diff
@@ -122,9 +122,9 @@ void Image::Shutdown() {
   memory_ = VK_NULL_HANDLE;
 }
 
-Result Image::CopyToHost(VkCommandBuffer command) {
+void Image::CopyToCPU(VkCommandBuffer command) {
```
[the remark this answers is not in the document]
Comment (dj2) [reply src/vulkan/image.cc]: I tend to think in terms of where is the memory accessible, We're making it available to the CPU

**Comment by dneto0** [comment]
Bots are failing too.

File: src/vulkan/pipeline.cc [hunk 14]
```diff
@@ -36,8 +37,8 @@ Pipeline::Pipeline(
     : device_(device),
       memory_properties_(properties),
       pipeline_type_(type),
-      fence_timeout_ms_(fence_timeout_ms),
-      shader_stage_info_(shader_stage_info) {}
+      shader_stage_info_(shader_stage_info),
+      fence_timeout_ms_(fence_timeout_ms) {}
```
Comment (dj2) [inline src/vulkan/pipeline.cc]: Why flip the order?

File: src/vulkan/pipeline.cc [hunk 15]
```diff
@@ -287,5 +288,40 @@ void Pipeline::BindVkPipeline() {
...
+
+  return Result("Vulkan::Pipeline descriptor with descriptor set: " +
+                std::to_string(descriptor_set) +
+                ", binding: " + std::to_string(binding) + " does not exist");
+}
+
+Result Pipeline::GetDescriptorInfo(const uint32_t descriptor_set,
+                                   const uint32_t binding,
+                                   ResourceInfo* info) {
+  for (size_t i = 0; i < descriptors_.size(); ++i) {
```
Comment (dj2) [inline src/vulkan/pipeline.cc]: assert(info)

File: src/vulkan/descriptor.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (dj2) [reply src/vulkan/descriptor.h]: Chatted with David about this and I was wrong. These should be Host/Device because there are cases where Host/Device is CPU/GPU and cases where it's CPU/CPU. So Host/Device is the more unique and correct naming.

Sorry about that.

File: src/verifier.cc (hunk 7 above)
Comment (jaebaek) [reply src/verifier.cc]: Sorry for making you confused. `size` here means the size of the ssbo in bytes. I don't know the number of elements at this moment. `cpu_memory` is just a pointer to ssbo. In my personal opinion, we would better not make a vector for this.

File: src/verifier.cc [hunk 16]
```diff
@@ -15,15 +15,73 @@
...
+        break;
+      case ProbeSSBOCommand::Comparator::kLessOrEqual:
+        if (*ptr > val)
+          return Result("Verifier failed: " + std::to_string(*ptr) + " > " +
+                        std::to_string(val));
+        break;
+      case ProbeSSBOCommand::Comparator::kGreater:
+        if (*ptr <= val)
+          return Result("Verifier failed: " + std::to_string(*ptr) +
+                        " <= " + std::to_string(val));
```
Comment (dj2) [inline src/verifier.cc]: For these error messages, I think we should use the comparator not the comparison used. So, this should be std::to_string(*ptr) + " > " + std::to_string(val); because that's the check in the script which failed.

File: src/verifier.cc [hunk 17]
```diff
@@ -15,15 +15,73 @@
...
+                                : static_cast<T>(v.AsDouble());
+
+    switch (comp) {
+      case ProbeSSBOCommand::Comparator::kEqual:
+        if (*ptr != val)
+          return Result("Verifier failed: " + std::to_string(*ptr) +
+                        " != " + std::to_string(val));
+        break;
+      case ProbeSSBOCommand::Comparator::kNotEqual:
+        if (*ptr == val)
```
Comment (dj2) [inline src/verifier.cc]: This and the != case probably need to be a fuzzy equal

File: src/vulkan/image.cc [hunk 18]
```diff
@@ -144,7 +140,7 @@ Result Image::CopyToHost(VkCommandBuffer command) {
                          GetHostAccessibleBuffer(), 1, &copy_region);
 
   MemoryBarrier(command);
-  return {};
+  return;
```
Comment (dj2) [inline src/vulkan/image.cc]: nit: return at end of method not needed.

**Comment by dneto0** [comment]
Please rebase.

File: src/vulkan/image.cc (hunk 18 above)
Comment (jaebaek) [reply src/vulkan/image.cc]: done

File: src/verifier.cc (hunk 16 above)
Comment (jaebaek) [reply src/verifier.cc]: done

File: src/dawn/pipeline_info.h [hunk 19]
```diff
@@ -50,6 +50,7 @@ struct RenderPipelineInfo {
   // The number of data bytes in the framebuffer host-side buffer.
   uint32_t fb_size = 0;
   bool fb_is_mapped = false;
+  const void* fb_data = false;
```
Comment (dj2) [inline src/dawn/pipeline_info.h]: This should be nullptr I think.

File: src/verifier.cc [hunk 20]
```diff
@@ -15,15 +15,95 @@
...
 
-const float kEpsilon = 0.002f;
+const double kEpsilon = 0.000001;
+const double kTolerance = 0.002;
 
-bool IsFloatPixelEqualInt(float pixel, uint8_t expected) {
+bool IsFuzzyEqual(const double real,
+                  const double expected,
+                  const double tolerance = kTolerance) {
   // TODO(jaebaek): Change kEpsilon to tolerance.
```
Comment (dj2) [inline src/verifier.cc]: This comment seems invalid now?

File: src/dawn/pipeline_info.h (hunk 19 above)
Comment (jaebaek) [reply src/dawn/pipeline_info.h]: done

File: src/verifier.cc (hunk 20 above)
Comment (jaebaek) [reply src/verifier.cc]: done

File: src/dawn/pipeline_info.h [hunk 21]
```diff
@@ -50,6 +50,7 @@ struct RenderPipelineInfo {
   // The number of data bytes in the framebuffer host-side buffer.
   uint32_t fb_size = 0;
   bool fb_is_mapped = false;
+  const void* fb_data = nullptr;
```
Comment (dneto0) [inline src/dawn/pipeline_info.h]: Every time fb_is_mapped is set, the fb_data member should be set.  For example where the buffer is unmapped.

File: src/dawn/engine_dawn.cc [hunk 22]
```diff
@@ -443,5 +430,28 @@ Result EngineDawn::CreateFramebufferIfNeeded() {
...
+
+  if (!render_pipeline_info_.fb_is_mapped)
+    return Result("Dawn: FrameBuffer is not mapped");
+
+  // TODO(dneto): Need to pass back a row_stride.
+  info->image_info.texel_stride = render_pipeline_info_.fb_texel_stride;
+  info->image_info.width = kFramebufferWidth;
+  info->image_info.height = kFramebufferHeight;
+  info->image_info.depth = 1U;
+  info->image_info.size_in_bytes = render_pipeline_info_.fb_texel_stride *
```
Comment (dneto0) [inline src/dawn/engine_dawn.cc]: This member doesn't exist in ResourceInfo.  I think you didn't compile this code.  :-)

File: src/engine.h [hunk 23]
```diff
@@ -104,11 +132,20 @@ class Engine {
...
-  virtual Result DoProcessCommands(uint32_t* stride,
-                                   uint32_t* width,
-                                   uint32_t* height,
-                                   const void** buf) = 0;
+  // Run all queued commands and copy frame buffer data to the host
+  // if graphics pipeline.
+  virtual Result DoProcessCommands() = 0;
+
+  // Get stride, width, height, and memory pointer of frame buffer.
+  virtual Result GetFrameBufferInfo(ResourceInfo* info) = 0;
```
Comment (dneto0) [inline src/engine.h]: Should say this is only valid if the framebuffer buffer is mapped into the host address space. In particular, if we have run DoProcessCommands() and that since then no graphics pipeline drawing commands have occurred (such as DoClear, DoDrawArrays, DoDrawRect).

File: src/dawn/pipeline_info.h (hunk 21 above)
Comment (dneto0) [reply src/dawn/pipeline_info.h]: I had another thought.   I want fb_data to be nullptr if and only if the buffer is *unmapped*.   But that makes fb_is_mapped redundant because it encodes the same information.
So: remove the fb_is_mapped variable entirely, and only set and check fb_data.

File: src/verifier.cc [hunk 24]
```diff
@@ -15,15 +15,80 @@
...
+#include "src/command.h"
 
 namespace amber {
 namespace {
 
-const float kEpsilon = 0.002f;
+const double kEpsilon = 0.000001;
+const double kTolerance = 0.002;
+
+bool IsFuzzyEqual(const double real,
```
[the remark this answers is not in the document]
Comment (dneto0) [reply src/verifier.cc]: I filed https://github.com/google/amber/issues/111  to sort out floating point comparisons better.

File: src/vulkan/buffer.cc [hunk 25]
```diff
@@ -71,6 +71,19 @@ void Buffer::CopyToDevice(VkCommandBuffer command) {
   MemoryBarrier(command);
 }
 
+void Buffer::CopyToHost(VkCommandBuffer command) {
+  if (is_buffer_host_accessible_)
```
Comment (dneto0) [inline src/vulkan/buffer.cc]: Even if the buffer is accessible to the host, you have to invalidate the mapped memory to ensure the host buffer gets the latest data.

https://www.khronos.org/registry/vulkan/specs/1.1/html/vkspec.html#vkInvalidateMappedMemoryRanges 

??

File: src/vulkan/buffer.h [hunk 26]
```diff
@@ -34,7 +34,8 @@ class Buffer : public Resource {
   Result CreateVkBufferView(VkFormat format);
   VkBufferView GetVkBufferView() const { return view_; }
 
-  // TODO(jaebaek): Determine copy all or partial data
+  // Only record the copy command to the command buffer and
```
Comment (dneto0) [inline src/vulkan/buffer.h]: Say what the method actually does.

File: src/vulkan/buffer.h [hunk 27]
```diff
@@ -45,6 +46,10 @@ class Buffer : public Resource {
     return Resource::GetHostAccessMemory();
   }
 
+  // Only record the copy command to the command buffer and
```
Comment (dneto0) [inline src/vulkan/buffer.h]: Say what the method actually does.

File: src/vulkan/compute_pipeline.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (dneto0) [reply src/vulkan/compute_pipeline.h]: TIL.

File: src/vulkan/engine_vulkan.cc [hunk 28]
```diff
@@ -272,28 +272,41 @@ Result EngineVulkan::DoPatchParameterVertices(
...
 
   auto graphics = pipeline_->AsGraphics();
-  Result r = graphics->ProcessCommands();
-
   auto frame = graphics->GetFrame();
-  *width = frame->GetWidth();
-  *height = frame->GetHeight();
-  *stride = VkFormatToByteSize(graphics->GetColorFormat());
-  *buf = frame->GetColorBufferPtr();
+  auto format_in_bytes = VkFormatToByteSize(graphics->GetColorFormat());
```
Comment (dneto0) [inline src/vulkan/engine_vulkan.cc]: bytes_per_texel

File: src/vulkan/image.h [hunk 29]
```diff
@@ -49,12 +46,13 @@ class Image : public Resource {
 
   // Resource
   VkDeviceMemory GetHostAccessMemory() const override {
-    if (is_image_host_accessible_)
-      return memory_;
-
     return Resource::GetHostAccessMemory();
   }
 
+  // Only record the copy command to the command buffer and
```
Comment (dneto0) [inline src/vulkan/image.h]: Say what the method actually does.

File: src/vulkan/resource.h [hunk 30]
```diff
@@ -31,10 +31,14 @@ class Resource {
     return host_accessible_memory_;
   }
 
+  virtual void CopyToHost(VkCommandBuffer command) = 0;
+
   virtual void Shutdown();
 
   void* HostAccessibleMemoryPtr() const { return memory_ptr_; }
 
+  size_t GetSize() const { return size_; }
```
Comment (dneto0) [inline src/vulkan/resource.h]: What are the units?  bytes?
What are you measuring? the whole underlying storage ?

File: src/verifier.cc [hunk 31]
```diff
@@ -64,6 +202,10 @@ Result Verifier::Probe(const ProbeCommand* command,
         std::to_string(frame_width) + "," + std::to_string(frame_height) + ")");
   }
 
+  double tolerance[4] = {};
+  bool is_tolerance_percent[4] = {};
+  SetupToleranceForTexels(command, tolerance, is_tolerance_percent);
```
Comment (dj2) [inline src/verifier.cc]: Do we need this? It seems like we could do something simple in the for loop below where if the tolerances size > 1 we take tolerance[i % 4] instead of building this array

File: src/verifier.cc [hunk 32]
```diff
@@ -15,15 +15,153 @@
...
+  if (is_tolerance_percent)
+    tolerance *= difference;
+  return difference < tolerance;
+}
 
-bool IsFloatPixelEqualInt(float pixel, uint8_t expected) {
-  // TODO(jaebaek): Change kEpsilon to tolerance.
-  return std::fabs(pixel - static_cast<float>(expected) / 255.0f) < kEpsilon;
+template <typename T>
+Result CheckValue(const ProbeSSBOCommand* command,
```
Comment (dj2) [inline src/verifier.cc]: Can these verifier changes be pulled out to a separate Cl and committed with the tests?

File: src/vkscript/executor_test.cc [hunk 33]
```diff
@@ -195,10 +195,11 @@ class EngineStub : public Engine {
     return {};
   }
 
-  Result DoProcessCommands(uint32_t*,
-                           uint32_t*,
-                           uint32_t*,
-                           const void**) override {
+  Result DoProcessCommands() override { return {}; }
```
Comment (dj2) [inline src/vkscript/executor_test.cc]: The change of DoProcessCommand into 3 methods could also possibly be split out to another CL

File: src/vulkan/storage_buffer_descriptor.cc [hunk 34]
```diff
@@ -119,6 +126,12 @@ Result StorageBufferDescriptor::UpdateDescriptorSet(
       descriptor_set, VK_DESCRIPTOR_TYPE_STORAGE_BUFFER, buffer_info);
 }
 
+void StorageBufferDescriptor::GetResourceInfo(ResourceInfo* info) {
```
Comment (dj2) [inline src/vulkan/storage_buffer_descriptor.cc]: Can this return a ResourceInfo instead of taking a pointer?

Comment (jaebaek) [reply src/vulkan/storage_buffer_descriptor.cc]: done

File: src/vulkan/buffer.cc [hunk 35]
```diff
@@ -65,12 +67,27 @@ void Buffer::CopyToDevice(VkCommandBuffer command) {
...
-  region.size = GetSize();
+  region.size = GetSizeInBytes();
 
   vkCmdCopyBuffer(command, GetHostAccessibleBuffer(), buffer_, 1, &region);
   MemoryBarrier(command);
 }
 
+Result Buffer::CopyToHost(VkCommandBuffer command) {
+  if (is_buffer_host_accessible_) {
+    return FlushAndInvalidateMemoryIfNeeded();
```
Comment (dneto0) [inline src/vulkan/buffer.cc]: It seems weird to put both flush and invalidate in the same call.
For copying to host you want to invalidate.
For copying to device from host, you want flush

File: src/vulkan/buffer.h [hunk 36]
```diff
@@ -34,7 +34,10 @@ class Buffer : public Resource {
   Result CreateVkBufferView(VkFormat format);
   VkBufferView GetVkBufferView() const { return view_; }
 
-  // TODO(jaebaek): Determine copy all or partial data
+  // If this buffer is host accessible, this method does nothing.
+  // Otherwise, this method only records the command for copying
+  // this buffer to the host accessible buffer. The actual
```
Comment (dneto0) [inline src/vulkan/buffer.h]: .. "this buffer to the host accessible buffer".  ??  Isn't that the wrong way around?

Comment (dneto0) [reply src/vulkan/buffer.h]: Also, compared to CopyToHost, why is this method different regarding buffers marked as COHERENT?

File: src/vulkan/resource.h [hunk 37]
```diff
@@ -68,6 +70,13 @@ class Resource {
            VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT;
   }
 
+  bool CheckMemoryHostCoherent(uint32_t memory_type_index) {
```
Comment (dneto0) [inline src/vulkan/resource.h]: "Check.." should be renamed to "Is...".

The "Check" is always ambiguous: i don't know what the true value means.  With "Is" it's always clear.

File: src/vulkan/buffer.cc (hunk 35 above)
Comment (jaebaek) [reply src/vulkan/buffer.cc]: Oh sorry. I was confused.

Updated.

File: src/vulkan/buffer.h (hunk 36 above)
Comment (jaebaek) [reply src/vulkan/buffer.h]: Updated.

File: src/vulkan/resource.h (hunk 37 above)
Comment (jaebaek) [reply src/vulkan/resource.h]: done

**Comment by dj2** [comment]
Can you rebase on top of ToT? The android and ndk-builds should work now.

**Comment by jaebaek** [comment]
Rebased
